### PR TITLE
Wi21 139 integrate with ibm

### PR DESCRIPTION
**Pull request by patryk-belka-wizzdev**

### CODE REVIEW

File: MicroPython/src/boot.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from uos import statvfs
 from gc import collect, enable, mem_alloc, mem_free, threshold
 from logging import DEBUG, basicConfig, debug, getLogger
-
+import micropython
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/boot.py:4]: Not needed

File: MicroPython/src/cloud/IBM_cloud.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import machine
+import ujson
+
+from common import config, utils
+from communication import wirerless_connection_controller
+from controller.main_controller_event import MainControllerEventType
+from cloud.cloud_interface import CloudProvider
+
+
+class IBMCloud(CloudProvider):
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/cloud/IBM_cloud.py:11]: Remove device_configuration from here

File: MicroPython/src/cloud/IBM_cloud.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    def __init__(self) -> None:
+        self.publish_success_topic = config.cfg.kaa_topic + '/status'
+        self.publish_error_topic = config.cfg.kaa_topic + '/error'
+
+    def device_configuration(self, data: list[dict]) -> int:
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/cloud/IBM_cloud.py:16]: Remove device_configuration from here

File: MicroPython/src/cloud/IBM_cloud.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                )
+                if ret:
+                    break
+            else:
+                logging.debug(
+                    "Tried to send data three times, failed! Aborting current measurement."
+                )
+
+        mqtt_communicator.disconnect()
+        wireless_controller.disconnect_station()
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/cloud/IBM_cloud.py:140]: Add returns (look at changed KAA class)

File: MicroPython/src/cloud/IBM_cloud.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        )
+
+        data = self._format_data(data)
+
+        logging.debug("data to send = {}".format(data))
+
+        ret = mqtt_communicator.publish_message(
+            payload=data, topic=config.cfg.ibm_topic, qos=config.cfg.QOS
+        )
+        if ret == False:
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/cloud/IBM_cloud.py:124]: change variable name to more readable one

File: MicroPython/src/common/config.py [hunk 6]
```diff
@@ -92,6 +92,25 @@
...
 DEFAULT_BLYNK_HUMIDITY_PIN = -1
 
+# IBM stuff
+IBM_CONFIG_PATH = "/resources/ibm_config.json"
+DEFAULT_IBM_DEVICE_ID = ""
+DEFAULT_IBM_USER = "use-token-auth"
+DEFAULT_IBM_ORGANIZATION_ID = ""
+DEFAULT_IBM_PASSWORD = ""
+DEFAULT_IBM_EVENT_ID =""
+DEFAULT_IBM_DEVICE_TYPE = "IoTStarter"
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/common/config.py:102]: remove hard coded default

File: MicroPython/src/common/config.py [hunk 7]
```diff
@@ -92,6 +92,25 @@
...
+DEFAULT_IBM_TOPIC = 'iot-2/evt/{}/fmt/json'.format(
+     DEFAULT_IBM_EVENT_ID
+ )
+#DEFAULT_IBM_TOPIC = 'iot-2/evt/topi/fmt/json'
+
+DEFAULT_IBM_HOST = '{}.messaging.internetofthings.ibmcloud.com'.format(DEFAULT_IBM_ORGANIZATION_ID)
+#'tf2veq.messaging.internetofthings.ibmcloud.com'
+#DEFAULT_IBM_CLIENT_ID = 'd:tf2veq:IoTStarter:IoTStarter-test'
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/common/config.py:110]: remove

Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/common/config.py:110]: remove commented lines

File: MicroPython/src/controller/main_controller.py [hunk 8]
```diff
@@ -3,6 +3,7 @@
 import machine
 import utime
 
+
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/controller/main_controller.py:6]: unnecessary line

File: MicroPython/src/data_upload/mqtt_communicator.py [hunk 9]
```diff
@@ -81,6 +81,21 @@ def __init__(self,
...
+        elif cloud_provider == Providers.IBM:
+            self.port = config.cfg.mqtt_port
+            self.server = config.cfg.ibm_host
+            self.client_id = config.cfg.ibm_client_id
+            self.username = config.cfg.ibm_user
+            self.password = config.cfg.ibm_password
+            self.MQTT_client = MQTTClient(
+                client_id=self.client_id,
+                server=self.server,
+                #server='tf2veq.messaging.internetofthings.ibmcloud.com',
```
Comment (jakub-pliszka-wizzdev) [inline MicroPython/src/data_upload/mqtt_communicator.py:93]: remove commented line

**Review by jakub-pliszka-wizzdev — APPROVED** [review]

File: IBM/README.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+[IBM Watson IoT Platform](https://internetofthings.ibmcloud.com/) is a clean and simple UI where you can simply and easily add and manage your devices, control access to your IoT service, and monitor your usage. It's free of charge with a limit of up to 500 devices and 200MB of data transfer.
+
+## Installation
+
+### Install requirements
+
+After you've created virtual environment, your current directory should be "iot-starter"
+
+``` 
+pip install pyserial cryptography click future pyelftools setuptools
```
Comment (damian-kurek-wizzdev) [inline IBM/README.md:17]: why this requirements aren't i requirements.txt file

File: IBM/README.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+Create an IoT Platform Service Lite instance directly from the [Platform Service Page in the IBM Cloud Service Catalog](https://cloud.ibm.com/catalog/services/internet-of-things-platform).  
+On the service page, click Launch to start administering your IBM Watson IoT Platform organization.
+The IBM Watson IoT Platform web console opens in a new browser tab at the following URL:
+https://<org_id>.internetofthings.ibmcloud.com/dashboard/#/overview Where org_id is the ID of your IBM Watson IoT Platform organization.
+
+### 3. Create a device
+
+1. In your IBM Watson IoT Platform dashboard, hover over the left side panel, and choose "Devices".
+2. Click "Add Device" in the top right corner.
+3. Create a device type. The device type name must be no more than 36 characters and contain only: Alpha-numeric characters, Hyphens(-), Underscores(_) and Periods(.). For example (IoTStarter)
```
Comment (damian-kurek-wizzdev) [inline IBM/README.md:41]: The device type name can be up to 36 characters long and contain

File: MicroPython/README.md [hunk 12]
```diff
@@ -97,6 +97,19 @@ By now, you should have three things:
...
+### IBM
+Make sure that your IBM Watson cloud is configured. For more information please go to README in "IBM" directory [here](../IBM/README.md).
+
+By now, you should have five things:
+ - Organization ID
+ - Device Type
+ - Device ID
+ - Authentication Method: "use-token-auth" by default
+ - Authentication Token
+
```
Comment (damian-kurek-wizzdev) [inline MicroPython/README.md:109]: random extra lines

File: MicroPython/src/cloud/IBM_cloud.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+            raise Exception("Create ibm_config.json file first!")
+
+        with open(config.IBM_CONFIG_PATH, "r", encoding="utf8") as infile:
+            config_dict = ujson.load(infile)
+
+        return config_dict
+
+    def _format_data(self, data: dict) -> dict:
+        """
+        Helper function for formatting data to match Kaa expected input
```
Comment (damian-kurek-wizzdev) [inline MicroPython/src/cloud/IBM_cloud.py:58]: KAA

**Review by damian-kurek-wizzdev — APPROVED** [review]

**Pull request merged**